File: functional-programming/Scala/jdoodle.scala
// turns out this is already functional, no changes nessacary!
//I just changed the testcases to match this project
object JDoodle {

    def encrypt(text: String, shift: Int): String = {
        text.map { // For each character in the text
            case char if char.isLetter && char.isLower =>
                val shiftedChar = ('a' + (char - 'a' + shift) % 26).toChar // Encrypt lowercase letters
                shiftedChar
            case char if char.isLetter && char.isUpper =>
                val shiftedChar = ('A' + (char - 'A' + shift) % 26).toChar // Encrypt uppercase letters
                shiftedChar
            case char => char // Preserve non-alphabetic characters
        }
        .mkString // Convert the encrypted characters back to a string
    }
    
    def decrypt(text: String, shift: Int): String = {
        text.map { // For each character in the text
            case char if char.isLetter && char.isLower =>
                val shiftedChar = ('a' + (char - 'a' - shift + 26) % 26).toChar // Decrypt lowercase letters
                shiftedChar
            case char if char.isLetter && char.isUpper =>
                val shiftedChar = ('A' + (char - 'A' - shift + 26) % 26).toChar // Decrypt uppercase letters
                shiftedChar
            case char => char // Preserve non-alphabetic characters
        }
        .mkString // Convert the decrypted characters back to a string
    }
    
    def solve(str: String, rotCount: Int = 25): List[String] = {
        (1 to rotCount).map(rot => decrypt(str, rot)).toList
    }
    
    def main(args: Array[String]): Unit = {
        
        println("Original: Pack my box with five dozen liquor jugs... then try to explain it to the TSA.")
        val encrypted = encrypt("Pack my box with five dozen liquor jugs... then try to explain it to the TSA.", 46)
        println(s"Encrypted: $encrypted")
        val decrypted = decrypt("Bmow yk naj iuft ruhq palqz xucgad vgse... ftqz fdk fa qjbxmuz uf fa ftq FEM.", 12)
        println(s"Decrypted: $decrypted")
        val cracked = solve("Bmow yk naj iuft ruhq palqz xucgad vgse... ftqz fdk fa qjbxmuz uf fa ftq FEM.", 25)
        println("Solved:")
        for ((crack, index) <-cracked.zipWithIndex){ //this is really cool, gets the index with the item as a pair
            println(s"rot: ${index + 1}: $crack")
        }
  }
}

File: functional-programming/Fsharp/jdoodle.fs
let encrypt (str : string) (rot : int) =
    str
    |> Seq.map (fun c ->
        match c with
        | c when c >= 'A' && c <= 'Z' -> char ((((int c - int 'A' + rot) % 26) + int 'A') : int)
        | c when c >= 'a' && c <= 'z' -> char ((((int c - int 'a' + rot) % 26) + int 'a') : int)
        | _ -> c)
    |> System.String.Concat

let decrypt (str : string) (rot : int) =
    str
    |> Seq.map (fun c ->
        match c with
        | c when c >= 'A' && c <= 'Z' -> char ((((int c - int 'A' - rot + 26) % 26) + int 'A') : int)
        | c when c >= 'a' && c <= 'z' -> char ((((int c - int 'a' - rot + 26) % 26) + int 'a') : int)
        | _ -> c)
    |> System.String.Concat

let solve (str : string) (rotCount : int) =
    [1..rotCount]
    |> List.map (fun rot -> decrypt str rot)

// Printouts
printfn "Original: Pack my box with five dozen liquor jugs... then try to explain it to the TSA."
printfn "Encrypted: %s" (encrypt "Pack my box with five dozen liquor jugs... then try to explain it to the TSA." 46)
printfn "Decrypted: %s" (decrypt "Bmow yk naj iuft ruhq palqz xucgad vgse... ftqz fdk fa qjbxmuz uf fa ftq FEM." 12)
let cracked = solve "Bmow yk naj iuft ruhq palqz xucgad vgse... ftqz fdk fa qjbxmuz uf fa ftq FEM." 25
printfn "Solved: "
cracked
|> List.iteri (fun i decrypted -> printfn "Rot: %d: %s" (i + 1) decrypted)


